app/services/db/monthly_sp_auth_count/unique_monthly_auth_counts_by_partner.rb

```diff
@@ -0,0 +1,144 @@
+# frozen_string_literal: true
+
+module Db
+  module MonthlySpAuthCount
+    module UniqueMonthlyAuthCountsByPartner
+      extend Reports::QueryHelpers
+
+      module_function
+
+      # @param [String] key label for billing (Partner requesting agency)
+      # @param [Array<String>] issuers issuers for the iaa
+      # @param [Date] start_date iaa start date
+      # @param [Date] end_date iaa end date
+      # @return [PG::Result, Array]
+      def call(key:, issuers:, start_date:, end_date:)
+        date_range = start_date...end_date
+
+        return [] if !date_range || issuers.blank?
+
+        # Query a month at a time, to keep query time/result size fairly reasonable
+        # The results are rows with [user_id, ial, auth_count, year_month]
+        months = Reports::MonthHelper.months(date_range)
+        queries = build_queries(issuers: issuers, months: months)
+
+        ial_to_year_month_to_users = Hash.new do |ial_h, ial_k|
+          ial_h[ial_k] = Hash.new { |ym_h, ym_k| ym_h[ym_k] = Multiset.new }
+        end
+
+        queries.each do |query|
+          temp_copy = ial_to_year_month_to_users.deep_dup
+
+          with_retries(
+            max_tries: 3,
+            rescue: [
+              ActiveRecord::SerializationFailure,
+              PG::ConnectionBad,
+              PG::TRSerializationFailure,
+              PG::UnableToSend,
+            ],
+            handler: proc do
+              ial_to_year_month_to_users = temp_copy
+              ActiveRecord::Base.connection.reconnect!
+            end,
+          ) do
+            Reports::BaseReport.transaction_with_timeout do
+              ActiveRecord::Base.connection.execute(query).each do |row|
+                user_id = row['user_id']
+                year_month = row['year_month']
+                ial = row['ial']
+                if row['min_returned_at'].nil? 
+                  next
+                elsif row['max_profile_verified_at'].nil?
+                  profile_age = -1
+                else
+                  profile_age = (Time.zone.parse(row['min_returned_at'].to_s).year - Time.zone.parse(row['max_profile_verified_at'].to_s).year)
```

Comment:
I am worried that doing `date1.year - date2.year` is too inaccurage for calculating profile age, because it does not ensure that 365 days have passed.

Example: these two dates are 1 day apart but area considered 1 year apart with this code
```ruby
date1 = Date.new(2023, 1, 1)
date2 = Date.new(2022, 12, 31)
(date1 - date2)
# 1
date1.year - date2.year
# 1
```

I think this is a reason to consider doing the date math inside the psql query:
```
SELECT DATE_PART('year', AGE('2023-12-31 00:00:00'::timestamp, '2022-12-31 00:00:00'::timestamp)) AS years_old;
 years_old 
-----------
         1
(1 row)
```

Reply:
This is a mechanism that we were not aware of. We wanted to do things in Ruby because it automatically factors in leap year on the correct year. The PSQL DATE_PART function I believe considers a year to be 365.25 days, which isn't really the same thing as an extra day every 4 years. If leap years are not a major concern we can switch it to calculate in PSQL since this error was overlooked by us

Reply:
That Ruby code as-is does not take leap years into account any better than the psql.

I think we can trust psql's 365.25 days/year approximation (ActiveSupport makes a similar assumption).

but we can also check with Billing Team and see if they would like to consider "1 year = 365 days, 2 years = 730 days" kind of logic and then we could use psql for that as well:

```
SELECT DATE_PART('day', '2023-12-31 00:00:00'::timestamp - '2022-12-31 00:00:00'::timestamp) AS days_old;
 days_old 
----------
      365
```

Reply:
After some more investigation it looks like PSQL does account for leap year accurately. I will move the profile age calculation to the SQL statement
`SELECT DATE_PART('day', '2023-03-01 00:00:00'::timestamp - '2023-02-01 00:00:00'::timestamp) AS days_old; days_old ----> 28
SELECT DATE_PART('day', '2024-03-01 00:00:00'::timestamp - '2024-02-01 00:00:00'::timestamp) AS days_old; days_old ----> 29`